src/protocol/prf_sharding/mod.rs

```diff
@@ -677,27 +679,33 @@ where
     F: PrimeField + ExtendableField,
 {
     let mut step: usize = 1 << BK::BITS;
-    let mut row_contribution = vec![value; 1 << BK::BITS];
+    let total_breakdowns = match breakdown_count {
+        Some(x) => x,
+        None => 1 << BK::BITS,
+    };
+    assert!(total_breakdowns <= 1 << BK::BITS);
+    let mut row_contribution = vec![value; total_breakdowns];
 
     for (tree_depth, bit_of_bdkey) in bd_key.iter().rev().enumerate() {
         let depth_c = ctx.narrow(&BinaryTreeDepthStep::from(tree_depth));
         let span = step >> 1;
-        let mut futures = Vec::with_capacity((1 << BK::BITS) / step);
-        for i in (0..1 << BK::BITS).step_by(step) {
+        let mut futures = Vec::with_capacity(bd_key.len() / step);
+        for i in (0..total_breakdowns).step_by(step) {
             let bit_c = depth_c.narrow(&BitOpStep::from(i));
 
-            if i + span < 1 << BK::BITS {
+            if i + span < total_breakdowns {
                 futures.push(row_contribution[i].multiply(bit_of_bdkey, bit_c, record_id));
             }
         }
         let contributions = ctx.parallel_join(futures).await?;
 
         for (index, bdbit_contribution) in contributions.into_iter().enumerate() {
             let left_index = index * step;
-            let right_index = left_index + span;
-
-            row_contribution[left_index] -= &bdbit_contribution;
-            row_contribution[right_index] = bdbit_contribution;
+            if left_index + span < total_breakdowns {
+                let right_index = left_index + span;
+                row_contribution[left_index] -= &bdbit_contribution;
```

Comment:
Shouldn't this be moved outside of the `if` statement? As in, shouldn't this always be done?

Reply:
I'm almost certain this is a bug. If the tests didn't catch this, we should add a test that would fail here until this is resolved.